packages/output/src/browser/output-widget.tsx

```diff
@@ -14,136 +14,244 @@
+        this.update();
+    }
+
+    protected onResize(message: Widget.ResizeMessage): void {
+        super.onResize(message);
+        this.update();
+    }
+
+    protected onUpdateRequest(message: Message): void {
+        super.onUpdateRequest(message);
+        if (!this.channel.isLocked) {
+            setTimeout(() => {
+                const anchor = this.anchor;
+                if (anchor) {
+                    anchor.scrollIntoView(false);
+                }
+            }, 1);
+        }
+    }
+
+    protected onActivateRequest(message: Message): void {
+        super.onActivateRequest(message);
+        const anchor = this.anchor;
+        if (anchor && !this.channel.isLocked) {
+            anchor.focus();
+        } else {
+            this.node.focus();
+        }
+    }
+
+    protected render(): React.ReactNode {
+        const lines = [];
+        let id = 0;
         const style: React.CSSProperties = {
             whiteSpace: 'pre',
             fontFamily: 'monospace',
         };
-
-        if (this.outputChannelManager.selectedChannel) {
-            for (const text of this.outputChannelManager.selectedChannel.getLines()) {
-                const lines = text.split(/[\n\r]+/);
-                for (const line of lines) {
-                    result.push(<div style={style} key={id++}>{line}</div>);
-                }
+        for (const text of this.channel.getLines()) {
+            for (const content of text.split(/[\n\r]+/) || []) {
+                lines.push(<div style={style} key={`${this.name}-${id++}`}>{content}</div>);
             }
         }
-        if (result.length === 0) {
-            result.push(<div style={style} key={id++}>{'<no output yet>'}</div>);
+        if (lines.length === 0) {
+            lines.push(<div style={style} key={`${this.name}-${id++}`}>{'<no output yet>'}</div>);
         }
-        return result;
+        const anchor = <div id={this.anchorId} tabIndex={0} />;
+        return <div>
+            {lines}
+            {anchor}
+        </div>;
     }
 
-    protected onUpdateRequest(msg: Message): void {
-        super.onUpdateRequest(msg);
-        setTimeout(() => {
-            const div = document.getElementById(OutputWidget.IDs.CONTENTS) as HTMLDivElement;
-            if (div && div.children.length > 0) {
-                div.children[div.children.length - 1].scrollIntoView(false);
-            }
-        });
+    get channel(): OutputChannel {
+        return this.options.channel;
     }
-}
 
-export namespace OutputWidget {
-    export namespace IDs {
-        export const CONTENTS = 'outputContents';
-        export const CHANNEL_LIST = 'outputChannelList';
+    get name(): string {
+        return this.channel.name;
     }
+
+    private get anchorId(): string {
+        return `output-channel-${this.name}-anchor`;
+    }
+
+    private get anchor(): HTMLElement | undefined {
+        const anchor = document.getElementById(this.anchorId);
```

Comment:
TODO: The React way `ref={this.setRef}` would be better, however, the original logic did it via `getElementById` lookup too.